src/main/java/com/google/firebase/auth/AuthProviderConfig.java

```diff
@@ -56,45 +62,52 @@ public boolean isEnabled() {
    * <p>Set the initial attributes of the new provider by calling various setter methods available
    * in this class.
    */
-  public abstract static class CreateRequest {
+  public abstract static class CreateRequest<T extends CreateRequest<T>> {
```

Comment:
Lets call this `AbstractCreateRequest` or `ProviderConfigCreateRequest` and reduce overusing the name `CreateRequest`.  This also doesn't need to be a nested class. Might be cleaner to put it elsewhere.

Reply:
I like the idea of calling it AbstractCreateRequest instead of CreateRequest. However, I don't think we should move this into a separate class. Keeping it together is more consistent with all of our other CreateRequest/UpdateRequest classes.

On a somewhat related note, should AuthProviderConfig be renamed to AbstractProviderConfig? (Of course, the decision here might be influenced by our generics decision in #400.)

Reply:
I prefer the shorter name `ProviderConfig` for this type (it's already nested in the `auth` package) so the "Auth" part is kind of redundant. Also may be this doesn't need to be an abstract class at all? Is there any behavior that child classes are expected to implement, or are they just adding more attributes to the parent type?

Reply:
I think renaming to `ProviderConfig` makes sense given that "Auth" is redundant.

However, I think we should keep it abstract for 2 reasons:
1) We never want to instantiate it on its own since it can't be used for anything.
2) If we make it concrete then we will be forced to implement `getThis` in `AuthProvider.AbstractCreateRequest`. The issue with this is that we then lose the ability to force its subclasses to implement `getThis`. If subclasses forget to override `getThis` then it would break how the chaining works.

Reply:
Does 2 actually apply here? Why would the outer class be required to implement a method of an abstract nested class?

Reply:
Sorry, you're correct. However, I think keeping it abstract makes it more clear that we do not intend on instantiating it on its own.